### PR TITLE
feat(sdk-core): tss ecdsa key creation flow with third party backup

### PR DESCRIPTION
**Pull request by pranavjain97**
## Description

Ticket: [BG-59055](https://bitgoinc.atlassian.net/browse/BG-59055)

This modifies the ecdsa TSS flow to support third party backup key
providers (for Nike, BitGo is the backup provider) where the backup key
is created via Bitgo (i.e WP->HSM). It involves two additional calls
to the WP, first to create the backup shares, and second time to
update the backup key in WP with the bitgoToBackup shares and verifying
the common keychain. Rest of the flow remains the same.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My code compiles correctly for both Node and Browser environments
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [x] The ticket or github issue was included in the commit message as a reference
- [x] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 1]
```diff
@@ -167,8 +268,102 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
+    }
+
+    const bitgoPublicGpgKey = await this.getBitgoPublicGpgKey();
+
+    const backupToUserNShare = await encryptNShareFromAPIKeyShare(backupToUserShare, userGpgKey.publicKey);
+    const encryptedNShares: DecryptableNShare[] = [
+      {
+        nShare: backupToUserNShare,
+        recipientPrivateArmor: userGpgKey.privateKey,
+        senderPublicArmor: bitgoPublicGpgKey.armor(),
```
Comment (pranavjain97) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:330]: Since backup is encrypted to the HSM, confirming this should be `bitgoPublicGpgKey.armor` and not user gpg public key, right?

File: modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export { getTxRequest, sendSignatureShare };
  * Combines YShares to combine the final TSS key
  * This can only be used to create the User or Backup key since it requires the common keychain from BitGo first
  *
- * @param params.keyShare - TSS key share
+ * @param params.userHeldKeyShare - TSS key share
```
Comment (LeoKotschenreuther) [inline modules/sdk-core/src/bitgo/tss/eddsa/eddsa.ts:30]: Why did you change this? In the function it's still called `keyShare`

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 3]
```diff
@@ -130,47 +133,116 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
   createUserKeychain(
     userGpgKey: SerializedKeyPair<string>,
     userKeyShare: KeyShare,
-    backupKeyShare: KeyShare,
+    backupKeyShare: BackupKeyShare,
     bitgoKeychain: Keychain,
     passphrase: string,
-    originalPasscodeEncryptionCode: string
+    originalPasscodeEncryptionCode?: string,
+    recipientIndex?: number,
```
Comment (LeoKotschenreuther) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:164]: This param is unused, right? If we don't need it then please remove it.

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 4]
```diff
@@ -196,8 +268,102 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
+  ): Promise<EncryptedNShare> {
+    let backupToBitgoShare: EncryptedNShare;
+    if (isThirdPartyBackup && backupShare.bitGoHeldKeyShares?.keyShares) {
+      const backupToOtherApiShare = backupShare.bitGoHeldKeyShares.keyShares.find(
+        (keyShare) => keyShare.from === 'backup' && keyShare.to === getParticipantFromIndex(recipientIndex)
+      );
+      if (!backupToOtherApiShare) {
+        throw new Error('Missing Backup to Bitgo key share');
+      }
+      backupToBitgoShare = await encryptNShareFromAPIKeyShare(backupToOtherApiShare, recipientGpgPublicArmor);
```
Comment (LeoKotschenreuther) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:288]: If we get the `backupToBitgoShare` already from BitGo then it should already be encrypted and we shouldn't need to do that in the SDK. In that case I'm expecting we pass it back to BitGo just the way we got it from BitGo.

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 5]
```diff
@@ -196,8 +268,102 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
+    }
+
+    const backupToUserShare = bitgoKeyShares.find((keyShare) => keyShare.from === 'backup' && keyShare.to === 'user');
+    if (!backupToUserShare) {
+      throw new Error('Missing BitGo to User key share');
+    }
+
+    const bitgoPublicGpgKey = await this.getBitgoPublicGpgKey();
+
+    const backupToUserNShare = await encryptNShareFromAPIKeyShare(backupToUserShare, userGpgKey.publicKey);
```
Comment (LeoKotschenreuther) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:325]: This should already be encrypted, right? I don't understand why we need to encrypt it here. I expect that the backup provider would take care of the encryption.

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts (hunk 4 above)
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:288]: Yup you're right. I will remove that encryption

**Comment by pranavjain97** [comment]
TODO: Need to add more unit / integration tests.

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 6]
```diff
@@ -510,17 +556,27 @@ describe('TSS Ecdsa Utils:', async function () {
...
 
+    const backupToUserPublicShare = Buffer.concat([
+      Buffer.from(backupKeyShare.nShares[1].y, 'hex'),
+      Buffer.from(backupKeyShare.nShares[1].chaincode, 'hex'),
+    ]).toString('hex');
+
+    const backupToBitgoPublicShare = Buffer.concat([
+      Buffer.from(backupKeyShare.nShares[3].y, 'hex'),
+      Buffer.from(backupKeyShare.nShares[3].chaincode, 'hex'),
+    ]).toString('hex');
```
Comment (pranavjain97) [inline modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:567]: This was failing later during user key combine step, because the public share needs to be the sum of the y value + chaincode.

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts [hunk 7]
```diff
@@ -510,17 +556,27 @@ describe('TSS Ecdsa Utils:', async function () {
       false,
     )];
 
+    const backupToUserPublicShare = Buffer.concat([
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:559]: Two reasons-
1. This is how the `chaincode` is being calculated. Without the concatentation, chaincode would be empty here. https://github.com/BitGo/BitGoJS/blob/dc9f5aef5804b6842205c2e0e50739660a8fc4ef/modules/sdk-core/src/bitgo/tss/ecdsa/ecdsa.ts#L371
2. Lot of other places the public share is y + chaincode. For eg. https://github.com/BitGo/BitGoJS/blob/f3342328a85c78ab9d886478bfd027239f2251d8/modules/sdk-core/src/bitgo/utils/tss/eddsa/eddsa.ts#L184

File: modules/sdk-core/src/bitgo/keychain/iKeychains.ts [hunk 8]
```diff
@@ -103,6 +104,7 @@ export interface CreateMpcOptions {
   passphrase?: string;
   originalPasscodeEncryptionCode?: string;
   enterprise?: string;
+  backupXpubProvider?: string;
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/keychain/iKeychains.ts:107]: Yeah i wasn't personally happy with this, but kept it for the sake of consistency. Now that I have a +1, going to change it.

File: modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (brekiy) [reply modules/bitgo/test/v2/unit/internal/tssUtils/ecdsa.ts:559]: Public/private share response types are [defined here](https://github.com/BitGo/bitgo-microservices/blob/d6e7d092b539e7124e6d1b1f929982d8e815e5bc/packages/wallet-platform/util/ksclient/types.ts#L236-L243) which may also be helpful.

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts [hunk 9]
```diff
@@ -92,33 +93,49 @@ export class EcdsaUtils extends baseTSSUtils<KeyShare> {
...
 
     const userKeyShare = await MPC.keyShare(1, m, n);
-    const backupKeyShare = await MPC.keyShare(2, m, n);
-
     const userGpgKey = await generateGPGKeyPair('secp256k1');
 
-    const bitgoKeychain = await this.createBitgoKeychain(userGpgKey, userKeyShare, backupKeyShare, params.enterprise);
-    const userKeychainPromise = this.createParticipantKeychain(
+    let isThirdPartyBackup = false;
+    // Currently, we only support bitGo as a third party backup provider for TSS
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsa.ts:106]: deez nits 😆

File: modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts [hunk 10]
```diff
@@ -59,30 +60,33 @@ export default class BaseTssUtils<KeyShare> extends MpcUtils implements ITssUtil
...
   }
 
   createBackupKeychain(
     userGpgKey: SerializedKeyPair<string>,
     userKeyShare: KeyShare,
-    backupKeyShare: KeyShare,
+    backupKeyShare: KeyShare | BackupKeyShare,
     bitgoKeychain: Keychain,
-    passphrase: string
+    passphrase: string,
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:77]: It should!

File: modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts [hunk 11]
```diff
@@ -59,30 +60,33 @@ export default class BaseTssUtils<KeyShare> extends MpcUtils implements ITssUtil
...
+    backupXpubProvider?: string,
+    isThirdPartyBackup?: boolean
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:79]: My motivation behind having both was that `isThirdPartyBackup` kinda verified earlier in the stack that the backupProvider is valid, but I think it makes more sense to have that check separately here.

[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/baseTSSUtils.ts:79]: created a helper to do the verification

**Comment by pranavjain97** [comment]
> TODO: Need to add more unit / integration tests.

Done ✅

File: modules/sdk-core/src/bitgo/utils/tss/baseTypes.ts [hunk 12]
```diff
@@ -204,6 +205,11 @@ export interface BitgoHeldBackupKeyShare {
   keyShares: ApiKeyShare[];
 }
 
+export interface BackupKeyShare {
```
[the remark this answers is not in the document]
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/utils/tss/baseTypes.ts:208]: I kinda like having a differentiation in types for the shares held by client vs bitgo, but yea I see that is not very scalable when we start supporting multiple KRS's. 

Also this might require a lot of work actually. Since the `userHeldKeyshare` is of type `KeyShare` which is accepted as is in many helper functions currently which work directly on the `pShares` and the `nShares`. 

Creating a new common type will require refactoring all those helpers, and i'm not sure if i have enough of the cryptographic context in the sdk to do that refactor.

**Review by olibeyene — APPROVED** [review]
lgtm

**Review by LeoKotschenreuther — APPROVED** [review]
